backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,358 @@
+                    if (newBackend == null) {
+                      logger.warn(String.format("Presently assigned backend=%s for process_group=%s is defunct but cannot find any available backend so keeping assignment unchanged",
+                          existingBackendAssociation, ProtoUtil.processGroupCompactRepr(processGroup)));
+                      future.complete(existingBackendAssociation);
+                    } else {
+                      try {
+                        deAssociateBackendWithProcessGroup(existingBackend, processGroup);
+                        associateBackendWithProcessGroup(newBackend, processGroup);
+                        if (logger.isDebugEnabled()) {
+                          logger.debug(String.format("process_group=%s, de-associating existing backend=%s, associating new backend=%s",
+                              ProtoUtil.processGroupCompactRepr(processGroup), existingBackend.getBackendIPAddress(), newBackend.getBackendIPAddress()));
+                        }
+                        future.complete(newBackend.getBackendIPAddress());
+                      } catch (Exception ex) {
+                        future.fail(String.format("Cannot persist association of backend=%s with process_group=%s in zookeeper",
+                            newBackend.getBackendIPAddress(), processGroup));
+                      }
+                    }
+                  } finally {
+                    if (newBackend != null) {
+                      try {
+                        availableBackendsByPriority.offer(newBackend);
+                      } catch (Exception ex) {
+                        //TODO: Some metric to indicate enqueue failure here
+                        logger.error("Error enqueuing backend=" + newBackend.getBackendIPAddress() + " back in the queue");
+                      }
+                    }
+                  }
+                }
+              }
+            } finally {
+              backendAssignmentLock.unlock();
+            }
+          } else {
+            future.fail("Timeout while acquiring lock on backend queue for process_group=" + processGroup);
+          }
+        } catch (InterruptedException ex) {
+          future.fail(new RuntimeException("Interrupted while acquiring lock on backend queue for process_group=" + processGroup, ex));
+        }
+      }, result.completer());
+    }
+
+    return result;
+  }
+
+  /**
+   * Finds an available backend from the queue. De-queues all backends which have become defunct since they were added
+   * @return available backend or null if none found
+   */
+  private BackendDetail getAvailableBackendFromPriorityQueue() {
+    BackendDetail availableBackend;
+    while((availableBackend = availableBackendsByPriority.poll()) != null) {
+      if(!availableBackend.isDefunct()) {
+        return availableBackend;
+      }
+    }
+    return (availableBackend == null || availableBackend.isDefunct()) ? null : availableBackend;
```

Comment:
Simplification: this statement should be "return null;" ? Because we come here iff either while loop wasn't entered, or we couldn't enter the return-condition.